reconcile/vault_replication.py

```diff
@@ -0,0 +1,176 @@
+import logging
+import re
+
+from reconcile.utils.vault import VaultClient, SecretNotFound
+from reconcile.gql_definitions.jenkins_configs import jenkins_configs
+
+from reconcile.gql_definitions.vault_policies import vault_policies
+from reconcile.gql_definitions.vault_instances import vault_instances
+from reconcile.gql_definitions.vault_instances.vault_instances import (
+    VaultReplicationV1,
+)
+from reconcile.gql_definitions.vault_policies.vault_policies import (
+    VaultPolicyV1,
+)
+
+from reconcile.utils import gql
+
+from typing import List
+
+QONTRACT_INTEGRATION = "vault-replication"
+
+
+def copy_vault_secret(
+    dry_run: bool, source_vault: VaultClient, dest_vault: VaultClient, path: str
+) -> None:
+
+    secret_dict = {"path": path, "version": "LATEST"}
+    secrets, version = source_vault.read_all_version(secret_dict)
+    for secret in secrets:
+        write_dict = {"path": path, "data": secret}
+        try:
+            _, dest_version = dest_vault.read_all(secret_dict)
+            if dest_version < version:
+                logging.info(
+                    [
+                        "replicate_vault_secret",
+                        "Secret found, new version",
+                        version,
+                        path,
+                    ]
+                )
+                if not dry_run:
+                    dest_vault.write(write_dict)
+            else:
+                logging.info(["replicate_vault_secret", dest_version, version, secret])
+        except SecretNotFound:
+            logging.info(["replicate_vault_secret", "Secret not found", path])
+            if not dry_run:
+                dest_vault.write(write_dict)
+
+
+def copy_vault_secrets(
+    dry_run: bool,
+    source_vault: VaultClient,
+    dest_vault: VaultClient,
+    path_list: List[str],
+    policy_paths: List[str],
+) -> None:
+
+    for path in path_list:
+        if is_valid_path(path, policy_paths):
```

Comment:
if I understand correctly this correct, this is going to silently discard the secrets it cannot copy, which is going to cause all kinds of misunderstandings... Most likely we want to collect all the secrets the policy doesn't allow us and fail or this is going to quickly blow on us.

Reply:
Solved on latest changes, now we fail with the complete list of secrets that can not be copied